File: back/OfferAPI/src/test/java/com/example/clientAPI/mapper/OfferMapperTest.java
package com.example.clientAPI.mapper;

import com.example.clientAPI.entity.OfferEntity;
import com.example.clientAPI.entity.OfferInputEntity;
import dto.offerapi.Offer;
import dto.offerapi.OfferInput;
import org.junit.jupiter.api.Test;

import java.time.LocalDate;
import java.util.Collections;
import java.util.List;

import static org.junit.jupiter.api.Assertions.*;

class OfferMapperTest {

    private final OfferMapper mapper = new OfferMapper();


    @Test
    void testToDTOFromEntity() {
        OfferEntity entity = new OfferEntity(1, "/images/test.jpg", "Titre",
                "Description", "active",
                LocalDate.of(2026, 1, 1), LocalDate.of(2026, 12, 31));

        Offer dto = mapper.toDTO(entity);

        assertNotNull(dto);
        assertEquals(1, dto.getId());
        assertEquals("/images/test.jpg", dto.getPicturePath());
        assertEquals("Titre", dto.getTitle());
        assertEquals("Description", dto.getDescription());
        assertEquals(Offer.StateEnum.ACTIVE, dto.getState());
        assertEquals(LocalDate.of(2026, 1, 1), dto.getStartDate());
        assertEquals(LocalDate.of(2026, 12, 31), dto.getEndDate());
    }

    @Test
    void testToDTONull() {
        assertNull(mapper.toDTO(null));
    }

    @Test
    void testToDTOInactiveState() {
        OfferEntity entity = new OfferEntity(2, null, "Titre", "Desc",
                "inactive", LocalDate.of(2026, 1, 1), LocalDate.of(2026, 6, 30));

        Offer dto = mapper.toDTO(entity);

        assertNotNull(dto);
        assertEquals(Offer.StateEnum.INACTIVE, dto.getState());
    }

    @Test
    void testToDTOExpiredState() {
        OfferEntity entity = new OfferEntity(3, null, "Titre", "Desc",
                "expired", LocalDate.of(2025, 1, 1), LocalDate.of(2025, 6, 30));

        Offer dto = mapper.toDTO(entity);

        assertNotNull(dto);
        assertEquals(Offer.StateEnum.EXPIRED, dto.getState());
    }


    @Test
    void testToEntityFromDTO() {
        Offer dto = new Offer();
        dto.setId(1);
        dto.setPicturePath("/images/test.jpg");
        dto.setTitle("Titre");
        dto.setDescription("Description");
        dto.setState(Offer.StateEnum.ACTIVE);
        dto.setStartDate(LocalDate.of(2026, 1, 1));
        dto.setEndDate(LocalDate.of(2026, 12, 31));

        OfferEntity entity = mapper.toEntity(dto);

        assertNotNull(entity);
        assertEquals(1, entity.getId());
        assertEquals("/images/test.jpg", entity.getPicturePath());
        assertEquals("Titre", entity.getTitle());
        assertEquals("Description", entity.getDescription());
        assertEquals("active", entity.getState());
        assertEquals(LocalDate.of(2026, 1, 1), entity.getStartDate());
        assertEquals(LocalDate.of(2026, 12, 31), entity.getEndDate());
    }

    @Test
    void testToEntityFromDTONull() {
        assertNull(mapper.toEntity((Offer) null));
    }


    @Test
    void testToDTOList() {
        OfferEntity e1 = new OfferEntity(1, null, "A", "D1", "active",
                LocalDate.of(2026, 1, 1), LocalDate.of(2026, 6, 30));
        OfferEntity e2 = new OfferEntity(2, null, "B", "D2", "inactive",
                LocalDate.of(2026, 3, 1), LocalDate.of(2026, 9, 30));

        List<Offer> dtos = mapper.toDTOList(List.of(e1, e2));

        assertEquals(2, dtos.size());
        assertEquals("A", dtos.get(0).getTitle());
        assertEquals("B", dtos.get(1).getTitle());
    }

    @Test
    void testToDTOListEmpty() {
        List<Offer> dtos = mapper.toDTOList(Collections.emptyList());
        assertTrue(dtos.isEmpty());
    }


    @Test
    void testToEntityList() {
        Offer d1 = new Offer();
        d1.setId(1);
        d1.setTitle("A");
        d1.setDescription("D1");
        d1.setState(Offer.StateEnum.ACTIVE);
        d1.setStartDate(LocalDate.of(2026, 1, 1));
        d1.setEndDate(LocalDate.of(2026, 6, 30));

        List<OfferEntity> entities = mapper.toEntityList(List.of(d1));

        assertEquals(1, entities.size());
        assertEquals("A", entities.get(0).getTitle());
    }

    @Test
    void testToEntityListEmpty() {
        List<OfferEntity> entities = mapper.toEntityList(Collections.emptyList());
        assertTrue(entities.isEmpty());
    }


    @Test
    void testToEntityFromOfferInput() {
        OfferInput input = new OfferInput();
        input.setPicturePath("/images/new.jpg");
        input.setTitle("Nouvelle Offre");
        input.setDescription("Nouvelle description");
        input.setState(OfferInput.StateEnum.ACTIVE);
        input.setStartDate(LocalDate.of(2026, 3, 1));
        input.setEndDate(LocalDate.of(2026, 6, 30));

        OfferInputEntity entity = mapper.toEntity(input);

        assertNotNull(entity);
        assertEquals("/images/new.jpg", entity.getPicturePath());
        assertEquals("Nouvelle Offre", entity.getTitle());
        assertEquals("Nouvelle description", entity.getDescription());
        assertEquals("active", entity.getState());
        assertEquals(LocalDate.of(2026, 3, 1), entity.getStartDate());
        assertEquals(LocalDate.of(2026, 6, 30), entity.getEndDate());
    }

    @Test
    void testToEntityFromOfferInputNull() {
        assertNull(mapper.toEntity((OfferInput) null));
    }


    @Test
    void testToInputDTO() {
        OfferInputEntity entity = new OfferInputEntity("/images/promo.jpg",
                "Promo", "Desc promo", "inactive",
                LocalDate.of(2026, 4, 1), LocalDate.of(2026, 4, 30));

        OfferInput dto = mapper.toInputDTO(entity);

        assertNotNull(dto);
        assertEquals("/images/promo.jpg", dto.getPicturePath());
        assertEquals("Promo", dto.getTitle());
        assertEquals("Desc promo", dto.getDescription());
        assertEquals(OfferInput.StateEnum.INACTIVE, dto.getState());
        assertEquals(LocalDate.of(2026, 4, 1), dto.getStartDate());
        assertEquals(LocalDate.of(2026, 4, 30), dto.getEndDate());
    }

    @Test
    void testToInputDTONull() {
        assertNull(mapper.toInputDTO(null));
    }
}

File: front/MyApplication/core/network/src/test/java/defalt/network/mapper/OfferMapperTest.kt
package defalt.network.mapper.offer

import defalt.domain.entity.offer.Offer as OfferEntity
import defalt.domain.entity.offer.OfferInput
import defalt.network.api.offer.model.Offer as OfferDto
import defalt.network.api.offer.model.OfferInput as OfferInputDto
import java.time.LocalDate
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNull
import org.junit.Test

class OfferMapperTest {


    private val today: LocalDate = LocalDate.parse("2025-01-01")
    private val nextMonth: LocalDate = LocalDate.parse("2025-02-01")



    @Test fun `OfferDto State ACTIVE to entity`() = assertEquals(OfferEntity.State.ACTIVE, OfferDto.State.ACTIVE.toEntity())

    @Test fun `OfferDto State INACTIVE to entity`() = assertEquals(OfferEntity.State.INACTIVE, OfferDto.State.INACTIVE.toEntity())

    @Test fun `OfferDto State EXPIRED to entity`() = assertEquals(OfferEntity.State.EXPIRED, OfferDto.State.EXPIRED.toEntity())

    @Test fun `OfferEntity State ACTIVE to dto`() = assertEquals(OfferDto.State.ACTIVE, OfferEntity.State.ACTIVE.toDto())

    @Test fun `OfferEntity State INACTIVE to dto`() = assertEquals(OfferDto.State.INACTIVE, OfferEntity.State.INACTIVE.toDto())

    @Test fun `OfferEntity State EXPIRED to dto`() = assertEquals(OfferDto.State.EXPIRED, OfferEntity.State.EXPIRED.toDto())

    @Test fun `roundtrip Offer State entity-dto-entity`() {
        OfferEntity.State.entries.forEach { e -> assertEquals(e, e.toDto().toEntity()) }
    }



    @Test fun `OfferDto toEntity maps all fields`() {
        val dto = OfferDto(
            id = 1,
            title = "Titre",
            description = "Desc",
            state = OfferDto.State.ACTIVE,
            startDate = today,
            endDate = nextMonth,
            picturePath = "/img.png",
        )
        val entity = dto.toEntity()
        assertEquals(1, entity.id)
        assertEquals("Titre", entity.title)
        assertEquals("Desc", entity.description)
        assertEquals(OfferEntity.State.ACTIVE, entity.state)
        assertEquals(today, entity.startDate)
        assertEquals(nextMonth, entity.endDate)
        assertEquals("/img.png", entity.picturePath)
    }

    @Test fun `OfferDto toEntity with null picturePath`() {
        val dto = OfferDto(
            id = 1,
            title = "T",
            description = "D",
            state = OfferDto.State.INACTIVE,
            startDate = today,
            endDate = nextMonth,
            picturePath = null,
        )
        assertNull(dto.toEntity().picturePath)
    }

    @Test fun `OfferEntity toDto maps all fields`() {
        val entity = OfferEntity(
            id = 2,
            title = "Test",
            description = "D",
            state = OfferEntity.State.EXPIRED,
            startDate = today,
            endDate = nextMonth,
        )
        val dto = entity.toDto()
        assertEquals(2, dto.id)
        assertEquals(OfferDto.State.EXPIRED, dto.state)
    }

    @Test fun `roundtrip Offer dto-entity-dto`() {
        val dto = OfferDto(
            id = 1,
            title = "T",
            description = "D",
            state = OfferDto.State.ACTIVE,
            startDate = today,
            endDate = nextMonth,
            picturePath = null,
        )
        assertEquals(dto, dto.toEntity().toDto())
    }

    @Test fun `List OfferDto toEntity`() {
        val list = listOf(
            OfferDto(id = 1, title = "A", description = "D", state = OfferDto.State.ACTIVE, startDate = today, endDate = nextMonth),
            OfferDto(id = 2, title = "B", description = "D", state = OfferDto.State.INACTIVE, startDate = today, endDate = nextMonth),
        )
        val entities = list.toEntity()
        assertEquals(2, entities.size)
        assertEquals(1, entities[0].id)
        assertEquals(2, entities[1].id)
    }



    @Test fun `OffersPostRequestEntity toDto maps all fields`() {
        val entity = OfferInput(
            title = "Titre",
            description = "Desc",
            state = OfferInput.State.ACTIVE,
            startDate = today,
            endDate = nextMonth,
            picturePath = null,
        )
        val dto = entity.toDto()

        assertEquals("Titre", dto.title)
        assertEquals(OfferInputDto.State.ACTIVE, dto.state)
        assertEquals(today, dto.startDate)
    }

    @Test fun `OffersPostRequest State mapping to OfferInput`() {

        assertEquals(OfferInputDto.State.ACTIVE, OfferInput.State.ACTIVE.toDto())
        assertEquals(OfferInputDto.State.INACTIVE, OfferInput.State.INACTIVE.toDto())
        assertEquals(OfferInputDto.State.EXPIRED, OfferInput.State.EXPIRED.toDto())
    }

    @Test fun `roundtrip OffersPostRequest entity-dto-entity removed`() {

    }



    @Test fun `OffersIdPutRequestEntity toDto maps all fields`() {
        val entity = OfferInput(
            title = "Nouveau",
            description = "Desc",
            state = OfferInput.State.ACTIVE,
            startDate = today,
            endDate = nextMonth,
            picturePath = null,
        )
        val dto = entity.toDto()
        assertEquals("Nouveau", dto.title)
        assertEquals(OfferInputDto.State.ACTIVE, dto.state)
    }

    @Test fun `OffersIdPutRequestEntity toDto with null state`() {
        val entity = OfferInput(
            title = "T",
            state = OfferInput.State.EXPIRED,
            description = "dsq",
            startDate = today,
            endDate = nextMonth,
            picturePath = "def",
        )

        assertEquals(OfferInputDto.State.EXPIRED, entity.toDto().state)
    }

    @Test fun `OffersIdPutRequest State mapping to OfferInput`() {
        assertEquals(OfferInputDto.State.ACTIVE, OfferInput.State.ACTIVE.toDto())
        assertEquals(OfferInputDto.State.INACTIVE, OfferInput.State.INACTIVE.toDto())
        assertEquals(OfferInputDto.State.EXPIRED, OfferInput.State.EXPIRED.toDto())
    }

    @Test fun `roundtrip OffersIdPutRequest entity-dto-entity removed`() {

    }




    @Test fun `OffersIdStatePatchRequestEntity ACTIVE toDto`() {
        val entity = OfferInput(
            state = OfferInput.State.ACTIVE,
            title = "fzfdz",
            description = "fdsfdf",
            startDate = today,
            endDate = nextMonth,
            picturePath = "fds",
        )
        assertEquals(OfferInputDto.State.ACTIVE, entity.toDto().state)
    }

    @Test fun `OffersIdStatePatchRequestEntity INACTIVE toDto`() {
        val entity = OfferInput(
            state = OfferInput.State.INACTIVE,
            title = "fzfdz",
            description = "fdsfdf",
            startDate = today,
            endDate = nextMonth,
            picturePath = "fds",
        )
        assertEquals(OfferInputDto.State.INACTIVE, entity.toDto().state)
    }

    @Test fun `OffersIdStatePatchRequestEntity EXPIRED toDto`() {
        val entity = OfferInput(
            state = OfferInput.State.EXPIRED,
            title = "fzfdz",
            description = "fdsfdf",
            startDate = today,
            endDate = nextMonth,
            picturePath = "fds",
        )
        assertEquals(OfferInputDto.State.EXPIRED, entity.toDto().state)
    }
}
